{"cells":[{"cell_type":"markdown","metadata":{"trusted":true},"source":["# Adding gnomAD and AlphaMissense information to VCF files\n","\n","##### Updated 06/03/2024\n","##### Selin Kubali\n","\n","##### Goal\n","Take cleaned files from *selected_genes/hcm/cleaned_vcf_files* and add gnomAD data and AlphaMissense scores. Filter for rare variants and exclude low-quality variants. Extrapolate unknown information to create a list of rare variants for each gene with predicted conservation and pathogenicity. Combine variant file with list of patients and demographic information to create a file for each gene that has all patients in UK BioBank, their demographic information, and any relevant information about variants they may carry.\n","\n","##### Input\n","Cleaned VCF files from *selected_genes/hcm/cleaned_vcf_files*\n","\n","CSV files with AlphaMissense scores, found in *selected_genes/hcm/vcf_files/carrier*\n","\n","VCF files with gnomAD information, found in *selected_genes/hcm/gnomAD/gnomAD_csvs*\n","\n","CSV files with patient carrier information, found in *selected_genes/hcm/csv_files*\n","\n","CSV file with patient demographic information, found in *selected_genes/hcm/csv_files*\n","\n","CSV file with polygenic risk scores for HCM for each patient, found in *selected_genes/hcm/csv_files*\n","\n","\n","##### Output\n","CSV for each gene with list of demographic and variant information for each patient, uploaded to *selected_genes/hcm/cox_model_data*\n","Variant information for each gene, uploaded to *selected_genes/hcm/variant_files*\n","___"]},{"cell_type":"code","execution_count":1,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["# load packages\n","import pandas as pd\n","from pandas.api.types import CategoricalDtype\n","from random import shuffle"]},{"cell_type":"code","execution_count":2,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def load_variants(gene, dir_path = \"selected_genes/hcm/cleaned_vcf_files\"):\n","    path = f'/mnt/project/{dir_path}/{gene}_variants_cleaned.vcf'\n","    variants = pd.read_csv(path, sep = \"\\t\")\n","    variants = variants.rename(columns = {'#CHROM':'Chrom', 'POS':'Pos','REF':'Ref','ALT':'Alt','ID':'Name','FILTER':'Filter'})\n","    variants = variants[variants['Filter'] == '.']\n","    variants = variants[['Chrom','Pos','Ref','Alt','Name']]\n","    return variants"]},{"cell_type":"code","execution_count":3,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def load_gnomAD(gene, dir_path = \"selected_genes/hcm/gnomAD/gnomAD_csvs\"):\n","    \"\"\"\" loads csv with gnomAD information \"\"\" \n","    path = f'/mnt/project/{dir_path}/{gene}_gnomAD.csv'\n","    gnomAD = pd.read_csv(path, sep = \",\")\n","    gnomAD['faf_max'] = gnomAD['fafmax_faf95_max']\n","    return gnomAD"]},{"cell_type":"code","execution_count":4,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def load_alphamissense(gene, dir_path = \"selected_genes/hcm/alpha_missensense_annotated\"):\n","    \"\"\"\" loads csv with alpha missense information \"\"\" \n","    path = f'/mnt/project/{dir_path}/{gene}_alphamissense.csv'\n","    alphamissense = pd.read_csv(path, sep = \",\")\n","    return alphamissense"]},{"cell_type":"markdown","metadata":{},"source":["### List of genes"]},{"cell_type":"code","execution_count":5,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\", \"TNNT2\", \"TPM1\", \"MYL2\", \"MYL3\", \"ACTC1\"]"]},{"cell_type":"code","execution_count":6,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for gene in genes:\n","    gnomAD = load_gnomAD(gene)\n","    variants = load_variants(gene)\n","    alphamissense = load_alphamissense(gene)\n","    gene_with_gnomAD = variants.merge(gnomAD, how = 'left', on = ['Chrom','Pos','Ref','Alt'])\n","    gene_with_gnomAD = gene_with_gnomAD.loc[(gene_with_gnomAD[\"faf_max\"] <= 0.001) | (gene_with_gnomAD[\"faf_max\"].isna())]     # filter for rare genes\n","    gene_gnomAD_ai = gene_with_gnomAD.merge(alphamissense, how = 'left', on = ['Chrom','Pos','Ref','Alt'])\n","    gene_gnomAD_ai = gene_gnomAD_ai.drop(['Unnamed: 0_x','Unnamed: 0_y','fafmax_faf95_max','Gene','Canonical','Source'],axis=1)\n","    gene_gnomAD_ai = gene_gnomAD_ai[gene_gnomAD_ai['Filter'] == 'PASS']     # filter for read quality\n","    gene_gnomAD_ai.to_csv(gene+\"_updated.csv\")"]},{"cell_type":"markdown","metadata":{},"source":["## Edit variants"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def comp_predictors_extreme_val(df, col):\n","    \"\"\" finds either lowest values or highest in a column for a computational score. returns this value \"\"\"\n","                    \n","    if col in ['am_pathogenicity', 'revel_max', 'cadd_raw_score', 'faf_max', 'phylop', 'spliceai_ds_max']:\n","        val = min(df[df[col].notna()][col])\n","        return val\n","                  "]},{"cell_type":"code","execution_count":7,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def define_comp_predictors(df, col):\n","    \"\"\" groups existing computation predictor scores (CADD_raw, phyloP100way_vertebrate, GERP++_RS, REVEL, am_pathogenicity) by functional consequences \n","    and assigns them to variants based on this consequence. \"\"\"\n","\n","\n","    for i in df.index:\n","        col_means_grouped = df.groupby(by='Consequence')[col].mean()\n","        if pd.isnull(df[col][i]):\n","            if (pd.isnull(col_means_grouped[df['Consequence'][i]])) & (df['Consequence'][i] == 'synonymous_variant'):\n","                if df[df[col].notna()][col].count() == 0:\n","                    df = df.drop([col], axis = 1)\n","                    break\n","               # what to do if no mean computational value exists for a Synonymous or Intronic consequence\n","                else:\n","                    df.loc[[i],[col]] = comp_predictors_extreme_val(df, col)\n","            else:\n","                df.loc[[i],[col]] = col_means_grouped[df['Consequence'][i]]\n","                \n","    return df"]},{"cell_type":"code","execution_count":10,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_variants(gene):\n","    \"\"\" load and clean variants df and fill in missing values \"\"\"\n","    variants = pd.read_csv(gene+'_updated.csv')\n","    variants = variants.drop(['Unnamed: 0'], axis = 1)\n","    \n","    # define deleterious mutations\n","    deleterious_list = [\"stop_gained\", \"start_lost\", \"splice_acceptor_variant\", \"splice_donor_variant\", \"splice_donor_region_variant\",\"splice_region_variant\",\"frameshift_variant\"]    \n","    deleterious = variants['Consequence'].apply(lambda x: any(substring in x for substring in deleterious_list))\n","    variants.loc[deleterious, 'Consequence'] = 'deleterious'\n","    variants = variants[variants['Consequence'].isin(['synonymous_variant','missense_variant','deleterious'])]\n","\n","    variants[\"val\"] = 1\n","\n","    # Extrapolate to fill in NA values \n","    variants = define_comp_predictors(variants, 'revel_max') \n","    variants = define_comp_predictors(variants, 'am_pathogenicity')   \n","    variants = define_comp_predictors(variants, 'cadd_raw_score') \n","    variants = define_comp_predictors(variants, 'faf_max') \n","    variants = define_comp_predictors(variants, 'phylop') \n","    variants = define_comp_predictors(variants, 'spliceai_ds_max') \n","\n","    variants = variants.drop(['val'], axis = 1)\n","\n","\n","\n","    return(variants)"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["genes = [\"ACTC1\", \"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"MYL2\", \"MYL3\",  \"PLN\", \"PTPN11\", \"TNNI3\", \"TNNT2\", \"TPM1\", \"TTR\"]\n","for g in genes:\n","    file_name = f'{g}_variants.csv'\n","    variants = get_variants(g)\n","\n","    variants.to_csv(file_name, index=False)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["!dx mkdir -p \"Cassa Lab Shared Project:selected_genes/hcm/variant_files/\"\n","!dx cd /selected_genes/hcm/variant_files/\n","!dx upload *_variants.csv"]},{"cell_type":"markdown","metadata":{},"source":["## All patients"]},{"cell_type":"code","execution_count":20,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_carriers(gene, dir_path=\"selected_genes/hcm/vcf_files/carrier\"):\n","    \"\"\"retrieve variant-patient mapping file\"\"\"\n","    \n","    path = f'/mnt/project/{dir_path}/{gene}.ssv'\n","    carriers = pd.read_csv(path,\n","                       sep=\" \",\n","                       names=[\"Chrom\", \"empty\", \"Pos\", \"Ref\", \"Alt\", \"na\", \"Carriers\"],\n","                       usecols=lambda x: x not in ['na', 'empty'] )\n","    \n","    carriers.loc[:, \"Carriers\"] = carriers[\"Carriers\"].str.strip(\"|\")\n","    carriers.loc[:, \"Carriers\"]= carriers[\"Carriers\"].str.split(\"|\", expand = False)\n","    carriers = carriers.explode(\"Carriers\").rename(columns={\"Carriers\": \"Carrier\"})\n","    carriers = carriers[carriers['Carrier'].apply(lambda x: not x.startswith('W'))] # remove withdrawn\n","    \n","    \n","    return(carriers)"]},{"cell_type":"code","execution_count":21,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_joined(gene, **paths):\n","    \"\"\"retrieves both variant and mapping files and joins them together by variant\"\"\"\n","    \n","    variants = get_variants(gene, paths[\"var_path\"]) if \"var_path\" in paths else get_variants(gene)\n","    carriers = get_carriers(gene, paths[\"car_path\"]) if \"car_path\" in paths else get_carriers(gene)\n","    variants = variants.astype({'Chrom': 'object', 'Pos':'object'})\n","    carriers = carriers.astype({'Pos':'object'})\n","    ann_carriers = carriers.merge(variants, on=[\"Chrom\", \"Pos\", \"Ref\", \"Alt\"], how = \"inner\")\n","    return(ann_carriers)\n"]},{"cell_type":"code","execution_count":22,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def get_most_severe(ann_carriers):\n","    \n","    \"\"\"accepts a filtered patient-variant df.\n","    selects and retains only the most severe variant per patient. returns this df.\"\"\"\n","    \n","    # order variant consequence by general severity (deleterious > missense > synonymous)\n","    consequence_cats = CategoricalDtype(categories=[\"synonymous_variant\", \"missense_variant\", \"deleterious\"], ordered=True)\n","    ann_carriers.loc[:, \"Consequence\"] = ann_carriers[\"Consequence\"].astype(consequence_cats)\n","\n","    # choose maximum AF for each variant, where possible\n","    ann_carriers.loc[:, \"max_AF\"] = ann_carriers.loc[:, [\"faf_max\"]].max(axis=1, skipna=True)\n","    \n","\n","    # create a random tie breaker index\n","    shuffled_index = list(range(0,len(ann_carriers)))\n","    shuffle(shuffled_index)\n","    ann_carriers.loc[:, \"tie_breaker\"] = shuffled_index\n","    \n","    # group by eid, sort by consequence > > confidence (for del. vars) > AF > tiebreaker\n","    # select top variant for each group\n","    ann_carriers_grouped = ann_carriers.groupby(\"Carrier\", group_keys=False).apply(pd.DataFrame.sort_values, [\"Consequence\", \"max_AF\", \"tie_breaker\"], ascending=[False, True, True])\n","    ann_carriers_grouped = ann_carriers_grouped.drop_duplicates(\"Carrier\", keep=\"first\")\n","    \n","    # important!! return categorical variables to normal objects so they don't clog pivot memory\n","    ann_carriers_grouped.loc[:, \"Consequence\"] = ann_carriers_grouped[\"Consequence\"].astype(object) \n","\n","    ann_carriers_grouped = ann_carriers_grouped.drop(['max_AF','tie_breaker'], axis = 1)\n","\n","    \n","    return(ann_carriers_grouped)"]},{"cell_type":"code","execution_count":24,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def patient_var_mappings(gene, most_severe=True, **paths):\n","    \"\"\"\n","    accepts gene name and some combination of var_path and car_path directories.\n","    returns a pandas df with one row per patient detailing their most severe variant\n","    \"\"\"\n","    annotated_carriers = get_joined(gene, **paths)\n","    \n","    if most_severe:\n","        annotated_carriers = get_most_severe(annotated_carriers)\n","        \n","    return annotated_carriers   "]},{"cell_type":"code","execution_count":25,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def merge_all_characteristics(gene, dir_path=\"selected_genes/hcm/csv_files\"):\n","    \n","    \"\"\" Add PRS and demographic information to each patient\"\"\"\n","    prs_csv = pd.read_csv(f'/mnt/project/{dir_path}/applied_hcm_prs.csv')\n","    phenotypic_info_csv = pd.read_csv(f'/mnt/project/{dir_path}/cardiomyopathy.csv')\n","    \n","    prs_csv['Carrier'] = prs_csv['eid'].astype(str)\n","    prs_csv = prs_csv.drop(['eid'], axis=1)\n","    phenotypic_info_csv['Carrier'] = phenotypic_info_csv['eid'].astype(str)  \n","    phenotypic_info_csv = phenotypic_info_csv.drop(['eid'], axis=1)\n","    \n","    g = patient_var_mappings(gene).reset_index()\n","    \n","    prs_joined = prs_csv.set_index('Carrier').join(g.set_index('Carrier'), how = 'outer')\n","    all_joined = phenotypic_info_csv.set_index('Carrier').join(prs_joined, how = 'outer')\n","    all_joined = all_joined.drop(['Pos', 'Ref', 'Alt', 'Chrom', 'index'], axis = 1)\n","\n","    return all_joined \n","    "]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["genes = [\"ACTC1\", \"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"MYL2\", \"MYL3\",  \"PLN\", \"PTPN11\", \"TNNI3\", \"TNNT2\", \"TPM1\", \"TTR\"]\n","for g in genes:\n","    file_name = f'{g}_with_generated_data.csv'\n","    all_carriers = merge_all_characteristics(g)\n","\n","    all_carriers.to_csv(file_name)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["!dx mkdir -p \"Cassa Lab Shared Project:selected_genes/hcm/cox_model_data/\"\n","!dx cd /selected_genes/hcm/cox_model_data/\n","!dx upload *_with_generated_data.csv"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":4}
