{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"name": "python", "version": "3.9.16", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# gnomAD VCF Extract\n##### Updated 06/03/2024\n##### Tian Yu, Lara Brown, Selin Kubali\n\n## Note:\nThis code is not meant to be run on JupyterLab and is provided here for convenience. Resources and explanation of how to run are provided at https://github.com/TYTYBU/vcfByGene\n\n#### Goal:\nExtract VCF files for exomic regions of given genes from gnomAD data.\n\n#### Required inputs\nSee https://github.com/TYTYBU/vcfByGene\n\n#### Output\nA VCF file containing information from gnomAD for each gene. Output to *selected_genes/hcm/gnomAD/gnomAD_gene_vcfs*\n", "metadata": {}, "id": "3ce035a1-89cc-4ea7-8a75-b97aee4de07f"}, {"cell_type": "markdown", "source": "### Load packages", "metadata": {}, "id": "5b47f712-9918-4e23-bf92-a811074c3f0d"}, {"cell_type": "code", "source": "from io import StringIO\nimport numpy as np\nimport pandas as pd\nfrom subprocess import call\nimport re", "metadata": {}, "execution_count": null, "outputs": [], "id": "5ab14e31-d796-48b8-ab05-5517bf85e470"}, {"cell_type": "markdown", "source": "### Set parameters", "metadata": {}, "id": "8edf64c6-81e8-412c-968d-115baeaa3be2"}, {"cell_type": "code", "source": "exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\nnumber_of_threads = 4 # number of threads used in bcftools output compression\ntag_str = 'skubali' # DNAnexus job tag\n\nproject_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n\ndx_vcf_path = project_path + \"selected_genes/hcm/gnomAD/\"\n\n\ndx_resource_path = project_path + \"GRCh38_resources/\"\ndx_script_path = project_path + \"scripts/\"\ndiff_bed = 'GRCh38_alldifficultregions.bed.gz'\nref_genome = 'GRCh38_reference_genome.fa'", "metadata": {"tags": []}, "execution_count": 19, "outputs": [], "id": "61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a"}, {"cell_type": "markdown", "source": "### Helper functions", "metadata": {}, "id": "f4000040-9b97-4085-a63e-b5aec384499f"}, {"cell_type": "markdown", "source": "### List of gene symbols as input", "metadata": {}, "id": "4ed62c94-9b34-42e2-90eb-67a527b04894"}, {"cell_type": "code", "source": "genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\", \"TNNT2\", \"TPM1\", \"MYL2\", \"MYL3\", \"ACTC1\", \"JPH2\", \"FHOD3\", \"CSRP3\", \"TRIM63\"]", "metadata": {}, "execution_count": 28, "outputs": [], "id": "8fbe087c-8e0b-4b71-9f81-d9fbada151d7"}, {"cell_type": "markdown", "source": "### Load MANE transcript coordinates", "metadata": {}, "id": "16fc2f8f-3742-4c4e-b01e-05d35069d52d"}, {"cell_type": "code", "source": "df = pd.read_csv(\"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\")\ndf = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\ndf = df[['seqname', 'start', 'end', 'gene_name']]\ndf['exon_flank_start'] = df['start'] - exon_flank_nt\ndf['exon_flank_end'] = df['end'] + exon_flank_nt\ndf['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\ndf", "metadata": {"tags": []}, "execution_count": 36, "outputs": [{"execution_count": 36, "output_type": "execute_result", "data": {"text/html": ["<div>\n", "<style scoped>\n", "    .dataframe tbody tr th:only-of-type {\n", "        vertical-align: middle;\n", "    }\n", "\n", "    .dataframe tbody tr th {\n", "        vertical-align: top;\n", "    }\n", "\n", "    .dataframe thead th {\n", "        text-align: right;\n", "    }\n", "</style>\n", "<table border=\"1\" class=\"dataframe\">\n", "  <thead>\n", "    <tr style=\"text-align: right;\">\n", "      <th></th>\n", "      <th>seqname</th>\n", "      <th>start</th>\n", "      <th>end</th>\n", "      <th>gene_name</th>\n", "      <th>exon_flank_start</th>\n", "      <th>exon_flank_end</th>\n", "      <th>region</th>\n", "      <th>pos_region</th>\n", "    </tr>\n", "  </thead>\n", "  <tbody>\n", "    <tr>\n", "      <th>21285</th>\n", "      <td>chr1</td>\n", "      <td>26067336</td>\n", "      <td>26067630</td>\n", "      <td>TRIM63</td>\n", "      <td>26067331</td>\n", "      <td>26067635</td>\n", "      <td>chr1:26067331-26067635</td>\n", "      <td>26067331-26067635</td>\n", "    </tr>\n", "    <tr>\n", "      <th>21288</th>\n", "      <td>chr1</td>\n", "      <td>26066268</td>\n", "      <td>26066440</td>\n", "      <td>TRIM63</td>\n", "      <td>26066263</td>\n", "      <td>26066445</td>\n", "      <td>chr1:26066263-26066445</td>\n", "      <td>26066263-26066445</td>\n", "    </tr>\n", "    <tr>\n", "      <th>21290</th>\n", "      <td>chr1</td>\n", "      <td>26061166</td>\n", "      <td>26061334</td>\n", "      <td>TRIM63</td>\n", "      <td>26061161</td>\n", "      <td>26061339</td>\n", "      <td>chr1:26061161-26061339</td>\n", "      <td>26061161-26061339</td>\n", "    </tr>\n", "    <tr>\n", "      <th>21292</th>\n", "      <td>chr1</td>\n", "      <td>26060266</td>\n", "      <td>26060361</td>\n", "      <td>TRIM63</td>\n", "      <td>26060261</td>\n", "      <td>26060366</td>\n", "      <td>chr1:26060261-26060366</td>\n", "      <td>26060261-26060366</td>\n", "    </tr>\n", "    <tr>\n", "      <th>21294</th>\n", "      <td>chr1</td>\n", "      <td>26058390</td>\n", "      <td>26058623</td>\n", "      <td>TRIM63</td>\n", "      <td>26058385</td>\n", "      <td>26058628</td>\n", "      <td>chr1:26058385-26058628</td>\n", "      <td>26058385-26058628</td>\n", "    </tr>\n", "    <tr>\n", "      <th>...</th>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "      <td>...</td>\n", "    </tr>\n", "    <tr>\n", "      <th>473557</th>\n", "      <td>chr20</td>\n", "      <td>44159618</td>\n", "      <td>44160407</td>\n", "      <td>JPH2</td>\n", "      <td>44159613</td>\n", "      <td>44160412</td>\n", "      <td>chr20:44159613-44160412</td>\n", "      <td>44159613-44160412</td>\n", "    </tr>\n", "    <tr>\n", "      <th>473559</th>\n", "      <td>chr20</td>\n", "      <td>44118505</td>\n", "      <td>44118623</td>\n", "      <td>JPH2</td>\n", "      <td>44118500</td>\n", "      <td>44118628</td>\n", "      <td>chr20:44118500-44118628</td>\n", "      <td>44118500-44118628</td>\n", "    </tr>\n", "    <tr>\n", "      <th>473561</th>\n", "      <td>chr20</td>\n", "      <td>44115665</td>\n", "      <td>44116386</td>\n", "      <td>JPH2</td>\n", "      <td>44115660</td>\n", "      <td>44116391</td>\n", "      <td>chr20:44115660-44116391</td>\n", "      <td>44115660-44116391</td>\n", "    </tr>\n", "    <tr>\n", "      <th>473563</th>\n", "      <td>chr20</td>\n", "      <td>44114782</td>\n", "      <td>44114876</td>\n", "      <td>JPH2</td>\n", "      <td>44114777</td>\n", "      <td>44114881</td>\n", "      <td>chr20:44114777-44114881</td>\n", "      <td>44114777-44114881</td>\n", "    </tr>\n", "    <tr>\n", "      <th>473566</th>\n", "      <td>chr20</td>\n", "      <td>44106590</td>\n", "      <td>44113503</td>\n", "      <td>JPH2</td>\n", "      <td>44106585</td>\n", "      <td>44113508</td>\n", "      <td>chr20:44106585-44113508</td>\n", "      <td>44106585-44113508</td>\n", "    </tr>\n", "  </tbody>\n", "</table>\n", "<p>334 rows \u00d7 8 columns</p>\n", "</div>"], "text/plain": ["       seqname     start       end gene_name  exon_flank_start  \\\n", "21285     chr1  26067336  26067630    TRIM63          26067331   \n", "21288     chr1  26066268  26066440    TRIM63          26066263   \n", "21290     chr1  26061166  26061334    TRIM63          26061161   \n", "21292     chr1  26060266  26060361    TRIM63          26060261   \n", "21294     chr1  26058390  26058623    TRIM63          26058385   \n", "...        ...       ...       ...       ...               ...   \n", "473557   chr20  44159618  44160407      JPH2          44159613   \n", "473559   chr20  44118505  44118623      JPH2          44118500   \n", "473561   chr20  44115665  44116386      JPH2          44115660   \n", "473563   chr20  44114782  44114876      JPH2          44114777   \n", "473566   chr20  44106590  44113503      JPH2          44106585   \n", "\n", "        exon_flank_end                   region         pos_region  \n", "21285         26067635   chr1:26067331-26067635  26067331-26067635  \n", "21288         26066445   chr1:26066263-26066445  26066263-26066445  \n", "21290         26061339   chr1:26061161-26061339  26061161-26061339  \n", "21292         26060366   chr1:26060261-26060366  26060261-26060366  \n", "21294         26058628   chr1:26058385-26058628  26058385-26058628  \n", "...                ...                      ...                ...  \n", "473557        44160412  chr20:44159613-44160412  44159613-44160412  \n", "473559        44118628  chr20:44118500-44118628  44118500-44118628  \n", "473561        44116391  chr20:44115660-44116391  44115660-44116391  \n", "473563        44114881  chr20:44114777-44114881  44114777-44114881  \n", "473566        44113508  chr20:44106585-44113508  44106585-44113508  \n", "\n", "[334 rows x 8 columns]"]}, "metadata": {}}], "id": "7c71254d-0da0-4458-92dd-3fcdf52e9b7b"}, {"cell_type": "code", "source": "genes_not_found = []\ngenes_found = []\nknown_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n\nfor gene in genes:\n    df_gene = df.loc[df['gene_name'] == gene]\n    if df_gene.shape[0] > 0:\n        genes_found.append(gene)\n        \n        region_str = ','.join(df_gene['region'].to_list())\n\n\n        match = re.match(r'^([^:]*)', region_str)\n        chrom = match.group(1)\n\n\n        name_input = \"gnomad.exomes.v4.1.sites.\"+chrom+\".vcf.bgz\"\n        name_output = \"vcf_for_\" + gene + \".gz\" \n\n        mem_level = \"mem1_ssd1_v2_x8\"\n        \n        bcf_cmd = \"bcftools view -r \" + region_str + \" -Oz -o\" + name_output + \" \" + name_input\n        dx_input_str = '-iin=' +'\"' +dx_vcf_path + name_input + '\"' + ' -iin=' + '\"' + dx_vcf_path + name_input + '.tbi' + '\"'\n        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcf_cmd + '\" --destination ' + dx_vcf_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n        !{dx_command}\n\n\n    else:\n        genes_not_found.append(gene)\n            \nprint('Genes not found in MANE database:')\nprint(genes_not_found)", "metadata": {"scrolled": true, "tags": []}, "execution_count": 13, "outputs": [{"name": "stdout", "output_type": "stream", "text": "chr1\n\njob-GkF01GjJqBj91VQb9ygbzKBX\n\nchr2\n\njob-GkF01J0JqBj72Gy0q0jKyXzK\n\nchr7\n\njob-GkF01J8JqBjBVypZ51BX68gQ\n\nchr11\n\njob-GkF01JQJqBj3g4FFqBYPvXbF\n\nchr6\n\njob-GkF01JjJqBj8pJb6p3X5XxPJ\n\nchr1\n\njob-GkF01K0JqBjK2V2VpyxvjJv5\n\nchr3\n\njob-GkF01K8JqBjK95GX6V746Qf6\n\nchr1\n\njob-GkF01KQJqBj4vbQ4J89vkJv7\n\nGenes not found in MANE database:\n\n[]\n"}], "id": "14ba3303-3410-4c56-a1f4-c6201572c1be"}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": [], "id": "0a2b673f-bcba-43c4-81db-09bfb57da0bc"}]}