{"cells":[{"cell_type":"markdown","metadata":{"tags":[]},"source":["# Loading gnomAD data\n","##### Updated 06/03/2024\n","##### Selin Kubali\n","\n","#### Goal:\n","Extract gnomAD information for all exomes in each given gene.\n","\n","#### Required inputs\n","- gnomAD VCFs for each gene. Found in *selected_genes/hcm/gnomAD/gnomAD_gene_vcfs*\n","\n","#### Output\n","CSV files containing gnomAD information – CADD, phyloP, allele frequency max, SpliceAI max, REVEL, Ensembl VEP consequence and confidence for canonical transcripts – for given genes. Uploaded to . Found in *selected_genes/hcm/gnomAD/gnomAD_csvs*."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import pandas as pd"]},{"cell_type":"markdown","metadata":{},"source":["## List of genes"]},{"cell_type":"code","execution_count":3,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\", \"TNNT2\", \"TPM1\", \"MYL2\", \"MYL3\", \"ACTC1\"]"]},{"cell_type":"markdown","metadata":{},"source":["## Download VCFs"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def download_vcfs(dir_path = 'selected_genes/hcm/gnomAD/gnomAD_gene_vcfs/'):\n","    !dx cd $dir_path\n","    !dx download vcf_*.gz\n","download_vcfs()"]},{"cell_type":"markdown","metadata":{},"source":["### Transform information given in gnomAD VCFs into a dataframe"]},{"cell_type":"code","execution_count":19,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def parse_string(info_str):\n","    \"\"\" Transform gnomAD data into a set of dictionaries \"\"\"\n","    features = info_str.split(';')\n","    features_list = []\n","    for pair in features:\n","        # Split once again on vep\n","        if 'vep=' in pair:\n","            vep_split = pair.split('vep=')\n","            try:\n","                features_list.append(['vep', vep_split[1]])\n","            except:\n","                print(vep_split)\n","        elif '=' in pair:\n","            features_list.append(pair.split('='))\n","\n","\n","\n","    features_dict = {key: value for key, value in features_list}\n","    filtered_features_dict = {key: features_dict.get(key, 'NA') for key in ['cadd_raw_score', 'phylop', 'fafmax_faf95_max', 'spliceai_ds_max', 'revel_max', 'vep']}\n","    return filtered_features_dict"]},{"cell_type":"code","execution_count":20,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def filter_list_of_lists(lst):\n","    \"\"\" Only add consequence for canonical transcripts from Ensembl VEP \"\"\"\n","    list_of_lists = []\n","    for i in lst:\n","        if len(i) > 32:\n","            if (i[24] == 'YES') & (i[32] == 'Ensembl'):\n","                list_of_lists.append([i[1], i[3], i[24], i[32], i[42]])\n","    return list_of_lists"]},{"cell_type":"code","execution_count":21,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def fill_na(row):\n","    \"\"\" Add row of NAs where information is missing \"\"\"\n","    if type(row) != list:\n","        return ['']*5\n","    else:\n","        return row"]},{"cell_type":"code","execution_count":22,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def clean_data(data):\n","    filtered_features_dict = data['Info'].apply(parse_string).tolist()\n","    # Convert provided dictionaries into dataframe \n","    info_df = pd.DataFrame(filtered_features_dict)\n","    chrom_data = pd.concat([filtered_features_dict, info_df], axis=1)\n","    # Split information given in VEP and filter for canonical transcripts\n","    chrom_data['vep'] = chrom_data['vep'].apply(lambda x: x.split(','))\n","    chrom_data['vep'] = chrom_data['vep'].apply(lambda x: [item.split('|') for item in x])\n","    chrom_data['vep'] = chrom_data['vep'].apply(filter_list_of_lists)\n","    chrom_data = chrom_data.explode('vep')\n","    chrom_data['vep'] = chrom_data['vep'].apply(fill_na)\n","    # Convert VEP data into multiple columns\n","    chrom_data[['Consequence', 'Gene', 'Canonical', 'Source', 'Confidence']] = pd.DataFrame(chrom_data['vep'].tolist(), index=chrom_data.index)\n","    chrom_data.drop('vep', axis=1, inplace=True)\n","    return chrom_data"]},{"cell_type":"code","execution_count":23,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for gene in genes:\n","    data = pd.read_csv('vcf_for_'+gene+'.gz', sep = '\\t', names = ['Chrom', 'Pos', 'ID', 'Ref', 'Alt', '?', 'Filter', 'Info'], usecols=['Chrom', 'Pos', 'Ref', 'Alt', 'Filter', 'Info'], low_memory=False,  comment='#', compression='gzip', memory_map=True)\n","    basic_data = data[['Chrom', 'Pos', 'Ref', 'Alt','Filter']]\n","    chrom_data = clean_data(data)\n","    chrom_data.to_csv(gene+'_gnomAD.csv')"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["!dx mkdir -p /selected_genes/hcm/gnomAD/gnomAD_csvs\n","!dx cd /selected_genes/hcm/gnomAD/gnomAD_csvs\n","!dx upload *_gnomAD.csv"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":4}
