{"cells":[{"cell_type":"code","execution_count":1,"id":"a9a7db64-5615-462a-adc3-446da0767246","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["import pandas as pd"]},{"cell_type":"markdown","id":"c15fed81-b6e5-418d-91fa-d90a53747cf7","metadata":{},"source":["# Loading AlphaMissense data\n","##### Updated 06/03/2024\n","##### Selin Kubali\n","\n","#### Goal:\n","Extract AlphaMissense scores for all exomes in each given gene.\n","\n","#### Required inputs\n","- MANE guide, found in *selected_genes/hcm/csv_files/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz*.\n","- AlphaMissense csv. Found in *selected_genes/hcm/csv_files/AlphaMissense_hg38.tsv\n","\n","#### Output\n","CSV files containing AlphaMissense scores for exomes in each gene. Found in *selected_genes/hcm/alpha_missensense_annotated*."]},{"cell_type":"markdown","id":"d5249368","metadata":{},"source":["##### Paths"]},{"cell_type":"code","execution_count":null,"id":"61f17b6c","metadata":{},"outputs":[],"source":["mane_path = f'/mnt/project/selected_genes/hcm/csv_files/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz'\n","alpha_missense_path = f'/mnt/project/selected_genes/hcm/csv_files/AlphaMissense_hg38.tsv'"]},{"cell_type":"code","execution_count":2,"id":"0ffd8c82-9ada-49c0-962d-f888f8601e5a","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["exon_flank_nt = 5"]},{"cell_type":"markdown","id":"4ed62c94-9b34-42e2-90eb-67a527b04894","metadata":{},"source":["### List of gene symbols as input"]},{"cell_type":"code","execution_count":3,"id":"8fbe087c-8e0b-4b71-9f81-d9fbada151d7","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["genes = [ \"ACTC1\", \"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"MYL2\", \"MYL3\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\", \"TNNT2\", \"TPM1\"]"]},{"cell_type":"markdown","id":"16fc2f8f-3742-4c4e-b01e-05d35069d52d","metadata":{},"source":["### Load MANE transcript coordinates"]},{"cell_type":"code","execution_count":null,"id":"c5fbf2c7","metadata":{},"outputs":[],"source":["df = pd.read_csv(mane_path)\n","df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n","df = df[['seqname', 'start', 'end', 'gene_name']]\n","df['exon_flank_start'] = df['start'] - exon_flank_nt\n","df['exon_flank_end'] = df['end'] + exon_flank_nt\n","df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n","df"]},{"cell_type":"code","execution_count":14,"id":"768b4f28-626b-4207-a24c-41e8d2d916b3","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["dtype_spec = {\n","\"Chrom\": str,\n","\"Pos\": int,\n","\"Ref\": str,\n","\"Alt\": str,\n","\"genome\": str,\n","\"uniprot_id\": str,\n","\"transcript_id\": str,\n","\"protein_variant\": str,\n","\"am_pathogenicity\": float,\n","\"am_class\": str\n","}\n","ai_scores = pd.read_csv(alpha_missense_path, sep = \"\\t\", skiprows = 4, names = [\"Chrom\", \"Pos\", \"Ref\", \"Alt\", \"genome\", \"uniprot_id\", \"transcript_id\", \"protein_variant\", \"am_pathogenicity\", \"am_class\"],usecols=[\"Chrom\", \"Pos\", \"Ref\", \"Alt\",\"am_pathogenicity\"])"]},{"cell_type":"code","execution_count":16,"id":"554a9303-ab36-43aa-9b4f-f9652cbd710f","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def in_ranges(pos, ranges):\n","    for start, end in ranges:\n","        if start <= pos <= end:\n","            return True\n","    return False"]},{"cell_type":"code","execution_count":17,"id":"275baa52-d0e7-4938-adbf-029611152688","metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for gene in genes:\n","    df_gene = df.loc[df['gene_name'] == gene]\n","    if df_gene.shape[0] > 0:\n","        \n","        combined = list(zip(df_gene['exon_flank_start'], df_gene['exon_flank_end']))\n","        ai_scores_gene = ai_scores[ai_scores['Pos'].apply(lambda x: in_ranges(x, combined))]\n","        ai_scores_gene.to_csv(gene+\"_alphamissense.csv\")"]},{"cell_type":"code","execution_count":null,"id":"25151f35-e033-46a9-8f96-f862194fb4e9","metadata":{},"outputs":[],"source":["!dx mkdir -p \"Cassa Lab Shared Project:/selected_genes/hcm/alpha_missensense_annotated\"\n","!dx cd /selected_genes/hcm/alpha_missensense_annotated\n","!dx upload *_alphamissense.csv"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":5}
