{"cells":[{"cell_type":"markdown","id":"d65b63c7-c164-4513-8241-e428e31f6eaa","metadata":{},"source":["# WES VCF Extract\n","##### Updated 06/03/2024\n","##### Tian Yu, Lara Brown\n","## Note:\n","This code is not meant to be run on DNANexus and is provided here for convenience. Resources and explanation of how to run are provided at https://github.com/TYTYBU/vcfByGene\n","\n","#### Goal:\n","Extract VCF files for exomic regions of given genes from UK BioBank WES data.\n","\n","#### Required inputs\n","See https://github.com/TYTYBU/vcfByGene\n","\n","#### Output\n","A VCF file for each gene. Output to */selected_genes/hcm/vcf_files*\n"]},{"cell_type":"code","execution_count":1,"id":"a9a7db64-5615-462a-adc3-446da0767246","metadata":{},"outputs":[],"source":["import pandas as pd"]},{"cell_type":"markdown","id":"8edf64c6-81e8-412c-968d-115baeaa3be2","metadata":{},"source":["### Set parameters"]},{"cell_type":"code","execution_count":2,"id":"61e00f50-ddf8-4ec3-9ba5-2ec86bc8361a","metadata":{"tags":[]},"outputs":[],"source":["exon_flank_nt = 5 # flanking nucleotides from the start and end of exons\n","number_of_threads = 4 # number of threads used in bcftools output compression\n","tag_str = 'skubali_hcm/vcf_files' # DNAnexus job tag\n","project_path = 'project-GGy3Bb0JqBj7zfxY8v4by61X:/'\n","# input vcf path (end with '/')\n","dx_vcf_path = project_path + \"Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ pVCF\\ format\\ -\\ final\\ release/\"\n","# output vcf path (end with '/')\n","dx_vcf_out_path = project_path + \"selected_genes/hcm/vcf_files\"\n","# resource path (end with '/')\n","dx_resource_path = project_path + \"GRCh38_resources/\"\n","# difficult regions bed filename\n","diff_bed = 'GRCh38_alldifficultregions.bed.gz'\n","# reference genome filename\n","ref_genome = 'GRCh38_reference_genome.fa' # index filename is inferred\n","mane_transcript = \"./resources/MANE.GRCh38.v1.0.select_ensembl_genomic.csv.gz\"\n","pvcf_block_path = \"./resources/pvcf_blocks.txt\"\n","\n","\n","!dx mkdir -p {dx_vcf_out_path} # create gene.vcf output folder"]},{"cell_type":"markdown","id":"f4000040-9b97-4085-a63e-b5aec384499f","metadata":{},"source":["### Helper functions"]},{"cell_type":"code","execution_count":3,"id":"2a8f096b-5b91-43aa-b546-4eaec039c30a","metadata":{},"outputs":[],"source":["def get_overlapping_UKB_vcfs(df_gene, df_blk, diagnosis=False):\n","    vcf_prefix = 'ukb23157_'\n","    vcf_suffix = '_v1.vcf.gz'\n","    \n","    all_vcf_files = []\n","    df_gene = df_gene.assign(vcf_files = '')\n","    for index, row in df_gene.iterrows():\n","        new_vcf_files = df_blk.loc[(df_blk['seqname'] == row['seqname']) & \n","                                   ((df_blk['end_pos'] >= row['exon_flank_start']) & (df_blk['start_pos'] <= row['exon_flank_end'])),\n","                                   'chr_blk_str'].tolist()\n","        new_vcf_files = list(vcf_prefix + x + vcf_suffix for x in set(new_vcf_files))\n","        df_gene.loc[index,'vcf_files'] = ','.join(new_vcf_files)\n","        all_vcf_files = all_vcf_files + new_vcf_files\n","    all_vcf_files = list(set(all_vcf_files))\n","    \n","    if diagnosis:\n","        return(df_gene)\n","    else:\n","        return(all_vcf_files)"]},{"cell_type":"markdown","id":"4ed62c94-9b34-42e2-90eb-67a527b04894","metadata":{},"source":["### List of gene symbols as input"]},{"cell_type":"code","execution_count":5,"id":"8fbe087c-8e0b-4b71-9f81-d9fbada151d7","metadata":{},"outputs":[],"source":["genes = [\"ACTN2\", \"ALPK3\", \"DES\", \"FLNC\", \"MYBPC3\", \"MYH6\", \"MYH7\", \"PLN\", \"PTPN11\", \"TNNI3\", \"TTR\", \"TNNT2\", \"TPM1\", \"MYL2\", \"MYL3\", \"ACTC1\"]"]},{"cell_type":"markdown","id":"214b5b7e-a208-4c88-bf56-19dbcf0034bd","metadata":{},"source":["### Load pVCF block coordinates"]},{"cell_type":"code","execution_count":6,"id":"62a230e4-d95d-4084-a914-01abecc3ff92","metadata":{"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ind</th>\n","      <th>chr</th>\n","      <th>blk</th>\n","      <th>start_pos</th>\n","      <th>end_pos</th>\n","      <th>seqname</th>\n","      <th>chr_blk_str</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>1218130</td>\n","      <td>chr1</td>\n","      <td>c1_b0</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>2</td>\n","      <td>1</td>\n","      <td>1</td>\n","      <td>1218131</td>\n","      <td>1426969</td>\n","      <td>chr1</td>\n","      <td>c1_b1</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>3</td>\n","      <td>1</td>\n","      <td>2</td>\n","      <td>1426970</td>\n","      <td>1758871</td>\n","      <td>chr1</td>\n","      <td>c1_b2</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>4</td>\n","      <td>1</td>\n","      <td>3</td>\n","      <td>1758872</td>\n","      <td>2514221</td>\n","      <td>chr1</td>\n","      <td>c1_b3</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>5</td>\n","      <td>1</td>\n","      <td>4</td>\n","      <td>2514222</td>\n","      <td>3782130</td>\n","      <td>chr1</td>\n","      <td>c1_b4</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>972</th>\n","      <td>973</td>\n","      <td>23</td>\n","      <td>20</td>\n","      <td>135552245</td>\n","      <td>141897932</td>\n","      <td>chrX</td>\n","      <td>cX_b20</td>\n","    </tr>\n","    <tr>\n","      <th>973</th>\n","      <td>974</td>\n","      <td>23</td>\n","      <td>21</td>\n","      <td>141897933</td>\n","      <td>152168662</td>\n","      <td>chrX</td>\n","      <td>cX_b21</td>\n","    </tr>\n","    <tr>\n","      <th>974</th>\n","      <td>975</td>\n","      <td>23</td>\n","      <td>22</td>\n","      <td>152168663</td>\n","      <td>153788223</td>\n","      <td>chrX</td>\n","      <td>cX_b22</td>\n","    </tr>\n","    <tr>\n","      <th>975</th>\n","      <td>976</td>\n","      <td>23</td>\n","      <td>23</td>\n","      <td>153788224</td>\n","      <td>156040895</td>\n","      <td>chrX</td>\n","      <td>cX_b23</td>\n","    </tr>\n","    <tr>\n","      <th>976</th>\n","      <td>977</td>\n","      <td>24</td>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>57227415</td>\n","      <td>chrY</td>\n","      <td>cY_b0</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>977 rows × 7 columns</p>\n","</div>"],"text/plain":["     ind  chr  blk  start_pos    end_pos seqname chr_blk_str\n","0      1    1    0          1    1218130    chr1       c1_b0\n","1      2    1    1    1218131    1426969    chr1       c1_b1\n","2      3    1    2    1426970    1758871    chr1       c1_b2\n","3      4    1    3    1758872    2514221    chr1       c1_b3\n","4      5    1    4    2514222    3782130    chr1       c1_b4\n","..   ...  ...  ...        ...        ...     ...         ...\n","972  973   23   20  135552245  141897932    chrX      cX_b20\n","973  974   23   21  141897933  152168662    chrX      cX_b21\n","974  975   23   22  152168663  153788223    chrX      cX_b22\n","975  976   23   23  153788224  156040895    chrX      cX_b23\n","976  977   24    0          1   57227415    chrY       cY_b0\n","\n","[977 rows x 7 columns]"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["df_blk = pd.read_table(pvcf_block_path, sep = '\\t', names = ['ind', 'chr', 'blk', 'start_pos', 'end_pos'])\n","df_blk['seqname'] = 'chr' + df_blk['chr'].map(str)\n","df_blk.loc[df_blk['seqname'] == 'chr23', 'seqname'] = 'chrX'\n","df_blk.loc[df_blk['seqname'] == 'chr24', 'seqname'] = 'chrY'\n","df_blk['chr_blk_str'] = df_blk['seqname'].str.replace('chr', 'c') + '_b' + df_blk['blk'].map(str)\n","df_blk"]},{"cell_type":"markdown","id":"16fc2f8f-3742-4c4e-b01e-05d35069d52d","metadata":{},"source":["### Load MANE transcript coordinates"]},{"cell_type":"code","execution_count":7,"id":"7c71254d-0da0-4458-92dd-3fcdf52e9b7b","metadata":{"tags":[]},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>seqname</th>\n","      <th>start</th>\n","      <th>end</th>\n","      <th>gene_name</th>\n","      <th>exon_flank_start</th>\n","      <th>exon_flank_end</th>\n","      <th>region</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>21285</th>\n","      <td>chr1</td>\n","      <td>26067336</td>\n","      <td>26067630</td>\n","      <td>TRIM63</td>\n","      <td>26067331</td>\n","      <td>26067635</td>\n","      <td>chr1:26067331-26067635</td>\n","    </tr>\n","    <tr>\n","      <th>21288</th>\n","      <td>chr1</td>\n","      <td>26066268</td>\n","      <td>26066440</td>\n","      <td>TRIM63</td>\n","      <td>26066263</td>\n","      <td>26066445</td>\n","      <td>chr1:26066263-26066445</td>\n","    </tr>\n","    <tr>\n","      <th>21290</th>\n","      <td>chr1</td>\n","      <td>26061166</td>\n","      <td>26061334</td>\n","      <td>TRIM63</td>\n","      <td>26061161</td>\n","      <td>26061339</td>\n","      <td>chr1:26061161-26061339</td>\n","    </tr>\n","    <tr>\n","      <th>21292</th>\n","      <td>chr1</td>\n","      <td>26060266</td>\n","      <td>26060361</td>\n","      <td>TRIM63</td>\n","      <td>26060261</td>\n","      <td>26060366</td>\n","      <td>chr1:26060261-26060366</td>\n","    </tr>\n","    <tr>\n","      <th>21294</th>\n","      <td>chr1</td>\n","      <td>26058390</td>\n","      <td>26058623</td>\n","      <td>TRIM63</td>\n","      <td>26058385</td>\n","      <td>26058628</td>\n","      <td>chr1:26058385-26058628</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>473557</th>\n","      <td>chr20</td>\n","      <td>44159618</td>\n","      <td>44160407</td>\n","      <td>JPH2</td>\n","      <td>44159613</td>\n","      <td>44160412</td>\n","      <td>chr20:44159613-44160412</td>\n","    </tr>\n","    <tr>\n","      <th>473559</th>\n","      <td>chr20</td>\n","      <td>44118505</td>\n","      <td>44118623</td>\n","      <td>JPH2</td>\n","      <td>44118500</td>\n","      <td>44118628</td>\n","      <td>chr20:44118500-44118628</td>\n","    </tr>\n","    <tr>\n","      <th>473561</th>\n","      <td>chr20</td>\n","      <td>44115665</td>\n","      <td>44116386</td>\n","      <td>JPH2</td>\n","      <td>44115660</td>\n","      <td>44116391</td>\n","      <td>chr20:44115660-44116391</td>\n","    </tr>\n","    <tr>\n","      <th>473563</th>\n","      <td>chr20</td>\n","      <td>44114782</td>\n","      <td>44114876</td>\n","      <td>JPH2</td>\n","      <td>44114777</td>\n","      <td>44114881</td>\n","      <td>chr20:44114777-44114881</td>\n","    </tr>\n","    <tr>\n","      <th>473566</th>\n","      <td>chr20</td>\n","      <td>44106590</td>\n","      <td>44113503</td>\n","      <td>JPH2</td>\n","      <td>44106585</td>\n","      <td>44113508</td>\n","      <td>chr20:44106585-44113508</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>334 rows × 7 columns</p>\n","</div>"],"text/plain":["       seqname     start       end gene_name  exon_flank_start  \\\n","21285     chr1  26067336  26067630    TRIM63          26067331   \n","21288     chr1  26066268  26066440    TRIM63          26066263   \n","21290     chr1  26061166  26061334    TRIM63          26061161   \n","21292     chr1  26060266  26060361    TRIM63          26060261   \n","21294     chr1  26058390  26058623    TRIM63          26058385   \n","...        ...       ...       ...       ...               ...   \n","473557   chr20  44159618  44160407      JPH2          44159613   \n","473559   chr20  44118505  44118623      JPH2          44118500   \n","473561   chr20  44115665  44116386      JPH2          44115660   \n","473563   chr20  44114782  44114876      JPH2          44114777   \n","473566   chr20  44106590  44113503      JPH2          44106585   \n","\n","        exon_flank_end                   region  \n","21285         26067635   chr1:26067331-26067635  \n","21288         26066445   chr1:26066263-26066445  \n","21290         26061339   chr1:26061161-26061339  \n","21292         26060366   chr1:26060261-26060366  \n","21294         26058628   chr1:26058385-26058628  \n","...                ...                      ...  \n","473557        44160412  chr20:44159613-44160412  \n","473559        44118628  chr20:44118500-44118628  \n","473561        44116391  chr20:44115660-44116391  \n","473563        44114881  chr20:44114777-44114881  \n","473566        44113508  chr20:44106585-44113508  \n","\n","[334 rows x 7 columns]"]},"execution_count":7,"metadata":{},"output_type":"execute_result"}],"source":["df = pd.read_csv(mane_transcript)\n","df = df.loc[(df['feature'] == 'exon') & (df['gene_name'].isin(genes))]\n","df = df[['seqname', 'start', 'end', 'gene_name']]\n","df['exon_flank_start'] = df['start'] - exon_flank_nt\n","df['exon_flank_end'] = df['end'] + exon_flank_nt\n","df['region'] = ((df['seqname'] + ':').str.cat(df['exon_flank_start'].astype(str)) + '-').str.cat(df['exon_flank_end'].astype(str))\n","df"]},{"cell_type":"markdown","id":"17e66005-ff41-4b75-a443-a2ed9df136bd","metadata":{},"source":["### Run Swiss-army-knife on DNAnexus\n","- get region info for each gene from block file\n","- bcftools command for step 2 finished\n","- bcftools command for step 3 & 4 TBD\n","- list genes not included in MANE set"]},{"cell_type":"code","execution_count":null,"id":"b03d87b8","metadata":{},"outputs":[],"source":["genes_not_found = []\n","genes_found = []\n","known_large_genes = [\"DSP\", \"TSC2\", \"TTN\", \"NCOA3\"]\n","\n","for gene in genes:\n","    df_gene = df.loc[df['gene_name'] == gene]\n","    if df_gene.shape[0] > 0:\n","        genes_found.append(gene)\n","        \n","        vcf_files = get_overlapping_UKB_vcfs(df_gene, df_blk)\n","        vcf_str = ' '.join(vcf_files)\n","        region_str = ','.join(df_gene['region'].to_list())\n","        \n","        if ((len(vcf_files) > 1) or (gene in known_large_genes)):\n","            mem_level = \"mem2_ssd1_v2_x16\" # dynamically change memory level when submitting jobs on DNAnexus\n","        else:\n","            mem_level = \"mem1_ssd1_v2_x4\"\n","        \n","        # filtering variants\n","        bcftools_cmd1 = \"bcftools concat -Ou -a -r \" + region_str + \" \" + vcf_str\n","        bcftools_cmd2 = \"bcftools view -Ou --max-alleles 5 -T ^\" + diff_bed\n","        bcftools_cmd3 = \"bcftools +fill-tags -Ou -- -t all\"\n","        bcftools_cmd4 = \"bcftools norm -Ou -m - -f \" + ref_genome \n","        bcftools_cmd5 = \"bcftools view -Oz -i 'AF<=0.001 && MAC >=1 && F_MISSING<0.1 && F_PASS(DP>=10 & GT!=\\\\\\\"mis\\\\\\\")> 0.9' > \" + gene + \"_variants.vcf.gz\"\n","        \n","        # extract carriers\n","        bcftools_cmd6 = \"mkdir -p carrier\"\n","        bcftools_cmd7 = \"bcftools query -i 'GT=\\\\\\\"RA\\\\\\\"|GT=\\\\\\\"AR\\\\\\\"|GT=\\\\\\\"AA\\\\\\\"' -f '%CHROM  %POS %REF %ALT %INFO/AF [%SAMPLE|]\\n' \" + gene + \"_variants.vcf.gz > carrier/\" + gene + \".ssv\"\n","        \n","        # parsing command\n","        bcftools_command_a = \" | \".join([bcftools_cmd1, bcftools_cmd2, bcftools_cmd3, bcftools_cmd4, bcftools_cmd5])\n","        bcftools_command_b = \" && \".join([bcftools_cmd6, bcftools_cmd7])\n","        bcftools_command = bcftools_command_a + \" && \" + bcftools_command_b\n","        \n","        # parsing input\n","        \n","        dx_input_str = ' '.join(set('-iin=\"' + dx_vcf_path + x + '\"' for x in vcf_files).union(set('-iin=\"' + dx_vcf_path + x + '.tbi\"' for x in vcf_files)))\n","\n","        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + diff_bed + '\"'\n","        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '\"'\n","        dx_input_str = dx_input_str + ' -iin=\"' + dx_resource_path + ref_genome + '.fai\"'\n","\n","        # final dx command\n","        dx_command = 'dx run swiss-army-knife --instance-type ' + mem_level + ' -y --brief ' + dx_input_str + ' -icmd=\"' + bcftools_command + '\" --destination ' + dx_vcf_out_path + ' --tag \"' + tag_str + '\" --property gene=' + gene\n","        !{dx_command}\n","    else:\n","        genes_not_found.append(gene)\n","            \n","print('Genes not found in MANE database:')\n","print(genes_not_found)"]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":5}
